### PR TITLE
Search filters: tax, expense type and tag

### PR DESCRIPTION
**Pull request by SzymczakJ**
### Details

This PR adds Expense Type, tax and Tag filters to Search feature. I decided to put these three filters together because they share the same concept and use the same main component. 
There are some files changed that don't belong to Search feature, but it's because I decided to make a small refactor of `PolicyTagList`, while implementing Tag filter(previous `PolicyTagList` really confused me).
I refactored Currency filter and Category filter to reuse common component.
I also refactored Search Pages to not use pointless `<FullPageNotFoundView shouldShow={false}>`
### Fixed Issues

$ https://github.com/Expensify/App/issues/46041
$ https://github.com/Expensify/App/issues/46037
$ https://github.com/Expensify/App/issues/46040
PROPOSAL:

### Tests

1. Go to advanced Filters page `/search/filters`
2. Apply tags filter(single value and then multiple values).
3. View results and check if they are filtered out.
4. Do the same for tax rates
5. Expense types are not yet testable because it hasn't got deployed yet.

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/c72bf5ad-ce81-4eaa-907b-3d918ae353da

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/9592fd06-4b52-4dab-9db6-e7173d4d79ab

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/643ca892-cda6-4464-8f4c-1be5da106e65

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/9a5efcbc-c7b2-4a85-b262-72527af19aaf

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/eda47708-3a3c-4465-9be7-56e49b8489b3

https://github.com/user-attachments/assets/ad5f0f94-e868-491a-aeae-2c3131a3463e

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/a06a0967-44e6-4610-8aa8-0fd5acec3a85

</details>

### CODE REVIEW

File: src/pages/Search/SearchFiltersExpenseTypePage.tsx
Comment (WojtekBoman) [inline src/pages/Search/SearchFiltersExpenseTypePage.tsx:50]: Do we have to to use `FullPageNotFoundView` if we always pass `false` here?

File: src/libs/SearchUtils.ts
Comment (WojtekBoman) [inline src/libs/SearchUtils.ts:466]: Maybe we can handle it in one if statement? Something like:

```
    if ([INPUT_IDS.TAX_RATE,  INPUT_IDS.EXPENSE_TYPE, INPUT_IDS.TAG].includes(filterKey) && filterValues[filterKey]) {
                const arrayFilterValues = filterValues[filterKey] ?? [];
                return `${CONST.SEARCH.SYNTAX_FILTER_KEYS[filterKey]}:${arrayFilterValues.map(sanitizeString).join(',')}`;
      }
```

**Comment by SzymczakJ** [comment]
I'll record videos tomorrow, but the code is ready for C+ review

File: src/CONST.ts
Comment (luacmartins) [inline src/CONST.ts:5343]: I wonder if we could just re-use `SYNTAX_FILTER_KEYS` and manipulate the values, e.g. `toUpperCase()` etc

**Comment by luacmartins** [comment]
Reassigning C+ as per https://expensify.slack.com/archives/C02NK2DQWUX/p1723051651463889

File: src/libs/PolicyUtils.ts
Comment (luacmartins) [inline src/libs/PolicyUtils.ts:292]:
```suggestion
function getTagNameFromTagLists(policyTagLists: PolicyTagLists): string[] {
```

**Review by luacmartins — COMMENTED** [review]
LGTM. I left a couple of small suggestions. I also spot checked some of the filters and they tested well. Gonna final approve once C+ is done reviewing.

**Review by rayane-d — CHANGES_REQUESTED** [review]

File: src/pages/Search/SearchFiltersTaxRatePage.tsx
Comment (luacmartins) [reply src/pages/Search/SearchFiltersTaxRatePage.tsx:33]: I think we should display the tax name in the filter and use the ID for the API call. We do similar things for the card filter for example

Comment (SzymczakJ) [reply src/pages/Search/SearchFiltersTaxRatePage.tsx:33]: I think that's a good idea, at first I opposed this one but when I thought it through.
In case when we have two workspaces and we add tax rate with the same name to both workspaces(they might have different values), The tax rate key will be the same 😄 . So basically it's the same as we would search over tax name.
These are two different workspaces with the tax rate with the same name:
Screenshot 2024-08-08 at 11 36 55
Screenshot 2024-08-08 at 11 36 15

Comment (SzymczakJ) [reply src/pages/Search/SearchFiltersTaxRatePage.tsx:33]: WDYT @luacmartins

File: src/components/SelectionList/SelectableListItem.tsx
Comment (SzymczakJ) [reply src/components/SelectionList/SelectableListItem.tsx:64]: It's the component designed to be used in Search feature and we don't have the need to show right element at the moment(It's only used in two places). So we decided to delete this. It was actually suggested by @Kicu.

File: src/libs/PolicyUtils.ts
Comment (SzymczakJ) [reply src/libs/PolicyUtils.ts:432]: Let's wait for response on this one https://github.com/Expensify/App/pull/46736#discussion_r1709076357.

File: src/libs/SearchUtils.ts
Comment (SzymczakJ) [reply src/libs/SearchUtils.ts:445]: Space seems reasonable, I'll add it.

File: src/libs/SearchUtils.ts
Comment (luacmartins) [reply src/libs/SearchUtils.ts:418]: Agree with @SzymczakJ

File: src/pages/Search/SearchFiltersTaxRatePage.tsx
Comment (luacmartins) [reply src/pages/Search/SearchFiltersTaxRatePage.tsx:33]: Looking into the backend part, but that makes sense to me

Comment (luacmartins) [reply src/pages/Search/SearchFiltersTaxRatePage.tsx:33]: I took a look at the backend and unfortunately we have no way to filter tax rates without the taxRateID, so we must send that. I think we could eliminate duplicated names like this (in a follow up PR):

1. Display de-duped names only
2. If more than one taxRateID for a given tax name exists, send both taxRateIDs to the API
3. The API will filter them by `WHERE taxRateID IN (<ids>)`

Comment (SzymczakJ) [reply src/pages/Search/SearchFiltersTaxRatePage.tsx:33]: I've thought this through and I think it's fine.
2. If more than one taxRateID for a given tax name exists, send both IDs to the API
this situation won't be possible because externalID is created based on taxRate name. So I just need to send one externalID.
Case closed I'll de-dup names and everything should work fine!

Comment (luacmartins) [reply src/pages/Search/SearchFiltersTaxRatePage.tsx]: I don't think we update the externalID if you rename the tax though, do we?

**Comment by SzymczakJ** [comment]
fixed your comments @rayane-djouah, can you continue reviewing? We still have a problem that Carlos stated in this [comment](https://github.com/Expensify/App/pull/46736#discussion_r1709702032) and I'll address it ASAP.
I also haven't solved Search header looking like this(id instead of name)
Screenshot 2024-08-08 at 17 16 32
but I think this should be addressed in another PR because this is a bigger problem that's also touching Card filters, keyword filters and From - To filters and would need to be a major refactor. @Kicu suggested solution that [here](https://github.com/Expensify/App/pull/46962#issue-2453345706)

**Review by rayane-d — CHANGES_REQUESTED** [review]

File: src/libs/SearchUtils.ts
Comment (luacmartins) [reply src/libs/SearchUtils.ts:448]: Quoted strings should be correctly parsed, unless there's a bug with the parser too :smil

**Comment by luacmartins** [comment]
@SzymczakJ is ooo today, so we'll have to wait until monday.

File: src/components/SearchMultipleSelectionPicker.tsx
Comment (Kicu) [inline src/components/SearchMultipleSelectionPicker.tsx:25]: @SzymczakJ can you please move this file inside `/components/Search`? this directory already exists and let's group all reusable Search related components there

**Review by Kicu — COMMENTED** [review]
added 1 comment about aligning dirs structure for Search

File: src/libs/SearchUtils.ts
Comment (luacmartins) [reply src/libs/SearchUtils.ts:448]: Sounds good

**Comment by luacmartins** [comment]
We still have comments to address and lint is failing

**Comment by SzymczakJ** [comment]
@rayane-djouah fixed comments and bugs that you mentioned. I'm recording videos at the moment.

**Comment by SzymczakJ** [comment]
Sorry for missing that, I've addressed that comment @rayane-djouah

**Comment by SzymczakJ** [comment]
Damn I fixed this but forgot to push it. Sorry for that blooper 😅

**Review by rayane-d — APPROVED** [review]
LGTM and tests well. 
@luacmartins, I think we can address https://github.com/Expensify/App/pull/46736#issuecomment-2284504935 and https://github.com/Expensify/App/pull/46736#issuecomment-2276105513 in follow-up PRs, as this PR is getting quite large.

**Comment by luacmartins** [comment]
I think https://github.com/Expensify/App/pull/46736#issuecomment-2284504935 is something we can put on the back burner for now. It's a bit of an edge case. I agree we can tackle the ID shown in the header in a follow up.

**Review by luacmartins — APPROVED** [review]
LGTM

**Comment by mvtglobally** [comment]
@luacmartins Do we need to QA anything here?

**Comment by luacmartins** [comment]
@mvtglobally not at this time.

**Comment by Beamanator** [comment]
FYI I believe this was deployed to prod yesterday, from this checklist - #47356

**Pull request merged**
